stdlib/in_channel.ml

```diff
@@ -63,4 +78,83 @@ let really_input_string ic len =
   | s -> Some s
   | exception End_of_file -> None
 
+(* Read up to [len] bytes into [buf], starting at [ofs]. Return total bytes
+   read. *)
+let read_upto ic buf ofs len =
+  let rec loop ofs len =
+    if len = 0 then ofs
+    else begin
+      let r = Stdlib.input ic buf ofs len in
+      if r = 0 then
+        ofs
+      else
+        loop (ofs + r) (len - r)
+    end
+  in
+  loop ofs len - ofs
+
+(* Return a buffer that has >= [ofs + n] bytes of storage, and
+   coincides with [buf] at indices < [ofs]. *)
+let ensure buf ofs n =
+  let len = Bytes.length buf in
+  if len >= ofs + n then buf
+  else begin
+    let new_len = ref len in
+    while !new_len < ofs + n do
+      new_len := 2 * !new_len + 1
+    done;
```

Comment:
Why the +1? We don't do that in buffer.ml

Reply:
Because `initial_size` may be 0.

Reply:
indeed